### PR TITLE
🐛  Add finalizer reconcile for Topology MachineSets and MachineDeployments

### PR DESCRIPTION
**Pull request by killianmuldoon**
Signed-off-by: killianmuldoon <kmuldoon@vmware.com>

Add reconciliation for the finalizer of MachineSets controlled by the topology controller. This ensures that if the finalizer is lost for any reason other than being removed by the topology/machineset controller it will be re-reconciled.

Fixes: #7530

### CODE REVIEW

File: internal/controllers/machineset/machineset_controller.go [hunk 1]
```diff
@@ -218,6 +220,18 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
...
+	// If the MachineSet is topology-managed ensure the MachineSetTopologyFinalizer is reconciled.
+	if feature.Gates.Enabled(feature.ClusterTopology) {
+		if _, ok := machineSet.Labels[clusterv1.ClusterTopologyOwnedLabel]; ok {
+			// - If the MachineSet is not being deleted.
+			if machineSet.DeletionTimestamp.IsZero() ||
+				// - If the MachineSet is being deleted but currently still has the Finalizer.
+				(!machineSet.DeletionTimestamp.IsZero() &&
+					controllerutil.ContainsFinalizer(machineSet, clusterv1.MachineSetTopologyFinalizer)) {
+				controllerutil.AddFinalizer(machineSet, clusterv1.MachineSetTopologyFinalizer)
+			}
```
Comment (fabriziopandini) [inline internal/controllers/machineset/machineset_controller.go:232]: Given that we are not using SSA, I will simplify this and just add the finalizer back if the machine is not being deleted

File: internal/controllers/machineset/machineset_controller.go [hunk 2]
```diff
@@ -218,6 +220,18 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
 		})
 	}
 
+	// If the MachineSet is topology-managed ensure the MachineSetTopologyFinalizer is reconciled.
```
Comment (fabriziopandini) [inline internal/controllers/machineset/machineset_controller.go:223]:
```suggestion
	// If the MachineSet is topology-managed ensure the MachineSetTopologyFinalizer is not removed before time.
```

File: internal/controllers/machineset/machineset_controller.go (hunk 1 above)
Comment (sbueringer) [reply internal/controllers/machineset/machineset_controller.go:232]: Sounds good to me

Comment (sbueringer) [reply internal/controllers/machineset/machineset_controller.go:232]: Given that this finalizer is actually owned by the MS **topology** controller. What about setting it in the MS topology controller?

Comment (vincepri) [reply internal/controllers/machineset/machineset_controller.go:232]: +1 to moving to the topology controller, it'd be consistent with the other PR as well #7531

File: internal/controllers/topology/machineset/machineset_controller.go [hunk 3]
```diff
@@ -119,12 +120,26 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
 	}
 
+	// Create a patch helper to add or remove the finalizer can be garbage collected by Kubernetes.
+	patchHelper, err := patch.NewHelper(ms, r.Client)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to create patch helper for %s", tlog.KObj{Obj: ms})
+	}
+	defer func() {
+		// Always attempt to patch the object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, ms); err != nil {
```
Comment (killianmuldoon) [inline internal/controllers/topology/machineset/machineset_controller.go:130]: Using the generic patch flow here, even though this controller should only ever try to patch metadata - adding or removing the finalizer.

File: internal/controllers/topology/machineset/machineset_controller.go [hunk 4]
```diff
@@ -119,12 +120,25 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
 	}
 
+	// Create a patch helper to add or remove the finalizer from the MachineSet.
+	patchHelper, err := patch.NewHelper(ms, r.Client)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to create patch helper for %s", tlog.KObj{Obj: ms})
+	}
+	defer func() {
+		if err := patchHelper.Patch(ctx, ms); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
```
Comment (sbueringer) [inline internal/controllers/topology/machineset/machineset_controller.go:130]:
```suggestion
			reterr = kerrors.NewAggregate([]error{reterr, errors.Wrapf(err, "failed to patch %s", tlog.KObj{Obj: ms})})
```
~

**Comment by sbueringer** [comment]
One nit, otherwise lgtm

Would be interesting to see if we can implement the MD finalizer thing in the same way on the other PR

**Comment by sbueringer** [comment]
(bug fix we want in both release series)

File: internal/controllers/topology/cluster/structuredmerge/options.go [hunk 5]
```diff
@@ -33,10 +33,11 @@ var (
 		{"metadata", "namespace"},
 		// uid is optional for a server side apply intent but sets the expectation of an object getting created or a specific one updated.
 		{"metadata", "uid"},
-		// the topology controller controls/has an opinion for labels, annotation, ownerReferences and spec only.
+		// the topology controller controls/has an opinion for labels, annotation, ownerReferences, finalizers and spec only.
 		{"metadata", "labels"},
 		{"metadata", "annotations"},
 		{"metadata", "ownerReferences"},
+		{"metadata", "finalizers"},
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/structuredmerge/options.go:40]: I would like to drop adding the finalizer in desired_state.go (l.626-631).
I think we shouldn't  add the finalizer on the first reconcile and then drop it in every subsequent one as we know have a stable solution in the MD topology controller.


No strong opinions on if we shoud leave finalizers here after dropping the code in desired_state.go. In other cases we only added paths to the list that we actually want to set somewhere.

Comment (sbueringer) [reply internal/controllers/topology/cluster/structuredmerge/options.go]: Let's wait to see what Fabrizio says

Comment (sbueringer) [reply internal/controllers/topology/cluster/structuredmerge/options.go]: EDIT: Also fine for me, let's see :)

Comment (killianmuldoon) [reply internal/controllers/topology/cluster/structuredmerge/options.go]: As discussed the new version only adds the finalizer in the MachineDeployment and MachineSet topology controllers. Previously the finalizer was still added on creation.

Comment (sbueringer) [reply internal/controllers/topology/cluster/structuredmerge/options.go]: Sounds good will take a look tomorrow

File: internal/controllers/topology/cluster/desired_state.go [hunk 6]
```diff
@@ -623,13 +622,6 @@ func computeMachineDeployment(_ context.Context, s *scope.Scope, desiredControlP
...
 	}
 
-	// If it's a new MachineDeployment, set the finalizer.
-	// Note: we only add it on creation to avoid race conditions later on when
-	// the MachineDeployment topology controller removes the finalizer.
-	if currentMachineDeployment == nil {
-		controllerutil.AddFinalizer(desiredMachineDeploymentObj, clusterv1.MachineDeploymentTopologyFinalizer)
-	}
-
 	// If an existing MachineDeployment is present, override the MachineDeployment generate name
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/desired_state.go:625]: Nit: please also drop the (buggy) assertion in cluster_controller_test.go (~ l.953-l.961) which tried to verify that the MD has the finalizer :)

**Comment by sbueringer** [comment]
Thank you very much!!

/lgtm

**Comment by fabriziopandini** [comment]
Great work nailing down this problem!
/lgtm
/approve

**Comment by sbueringer** [comment]
Thx!

(@Fabrizio we had to commit --amend push force to re-trigger EasyCLA, it was stuck somehow)

/lgtm

**Pull request merged**